plugins/gr-limesdr/lib/device_handler.cc

```diff
@@ -0,0 +1,753 @@
+                        "block_type value {:d}.",
+                        block_type));
+    }
+
+    // Check block settings which must match
+    if (device_vector[device_number].source_flag &&
+        device_vector[device_number].sink_flag) {
+        // Chip_mode must match in blocks with the same serial
+        if (device_vector[device_number].source_channel_mode !=
+            device_vector[device_number].sink_channel_mode) {
+
+            close_all_devices();
+            throw std::invalid_argument(fmt::format(
+                "device_handler::check_blocks(): channel mismatch in LimeSuite "
+                "Source (RX) ({:d}) and LimeSuite Sink (TX) ({:d})",
+                device_vector[device_number].source_channel_mode,
+                device_vector[device_number].sink_channel_mode));
+        }
+
+        // When file_switch is 1 check filename match throughout the blocks with the same
+        // serial
+        if (device_vector[device_number].source_filename !=
+            device_vector[device_number].sink_filename) {
+
+            close_all_devices();
+            throw std::invalid_argument(fmt::format(
+                "device_handler::check_blocks(): file must match in LimeSuite "
+                "Source (RX) ({:s}) and LimeSuite Sink (TX) ({:s})",
+                device_vector[device_number].source_filename,
+                device_vector[device_number].sink_filename));
+        }
+    }
+}
+
+void device_handler::settings_from_file(int device_number,
+                                        const std::string& filename,
+                                        int* pAntenna_tx)
+{
+    auto status =
+        device_handler::getInstance().get_device(device_number)->LoadConfig(0, filename);
+    if (status != lime::OpStatus::SUCCESS)
+        device_handler::getInstance().error(device_number);
+
+    // Set LimeSDR-Mini switches based on .ini file
+    std::array<int, 2> antenna_rx = { 0, 0 };
+    std::array<int, 2> antenna_tx = { 0, 0 };
+    antenna_tx[0] = device_handler::getInstance()
+                        .get_device(device_number)
+                        ->GetAntenna(0, lime::TRXDir::Tx, 0);
+    /* Don't print error message for the mini board */
+    suppress_limesuite_logging();
+    antenna_tx[1] = device_handler::getInstance()
+                        .get_device(device_number)
+                        ->GetAntenna(0, lime::TRXDir::Tx, 1);
+    antenna_rx[1] = device_handler::getInstance()
+                        .get_device(device_number)
+                        ->GetAntenna(0, lime::TRXDir::Rx, 1);
+
+    // Restore our logging
+    set_limesuite_logger();
+    antenna_rx[0] = device_handler::getInstance()
+                        .get_device(device_number)
+                        ->GetAntenna(0, lime::TRXDir::Rx, 0);
+
+    if (pAntenna_tx != nullptr) {
+        pAntenna_tx[0] = antenna_tx[0];
+        pAntenna_tx[1] = antenna_tx[1];
+    }
+
+    device_handler::getInstance()
```

Comment:
why not just store "device_handler::getInstance()" into local variable once at the start of the function, and then reuse it? Same with `device_handler::getInstance().get_device(device_number)`, it would greatly reduce spam and improve readablility.